docs/tutorials/quickstart.rst

```diff
@@ -0,0 +1,109 @@
+* `VirtualBox <https://www.virtualbox.org/>`_: a software virtualization tool
+* `Vagrant <https://www.vagrantup.com/downloads.html>`_: a command line utility for managing the lifecycle of virtual machines
+
+As the focus of this tutorial is NAPALM, we don't even scratch the surface of these tools.  If you're not familiar with them, please do some research [#f1]_ as they will be an important part of your development/ops toolkit.
+
+Installation
+------------
+
+Install NAPALM with pip::
+
+    pip install napalm
+
+Arista vEOS
+-----------
+
+We'll use Arista, which can be downloaded for free from the Arista site.
+
+Create an account at https://www.arista.com/en/user-registration, and go to https://www.arista.com/en/support/software-download.
+
+Download the latest "vEOS-lab-<version>-virtualbox.box" listed in the vEOS folder at the bottom of the page.
+
+Add it to your vagrant box list, changing the `<version>`::
+
+    $ vagrant box add --name vEOS-lab-<version>-virtualbox ~/Downloads/vEOS-lab-<version>-virtualbox.box
+    $ vagrant box list
+    vEOS-lab-quickstart (virtualbox, 0)
+
+You can delete the downloaded .box file once you have added it, as ``vagrant box add`` copies downloaded file to a designated directory (e.g., for Mac OS X and Linux: ``~/.vagrant.d/boxes``, Windows: ``C:/Users/USERNAME/.vagrant.d/boxes``).
+
+Starting Vagrant
+----------------
+
+Create a Vagrantfile on your machine with the following content:
+
+.. literalinclude:: Vagrantfile
+   :language: ruby
+
+The above content is also available on `GitHub <https://raw.githubusercontent.com/napalm-automation/napalm/master/docs/tutorials/Vagrantfile>`_.
+
+This Vagrantfile creates a base box and a vEOS box when you call "vagrant up"::
+
+    $ vagrant up
+    ... [output omitted] ...
+
+    $ vagrant status
+    Current machine states:
+    base                      running (virtualbox)
+    eos                       running (virtualbox)
+
+You may see some errors when the eos box is getting created [#f2]_.
+
+
+Using the NAPALM command-line client
+------------------------------------
+
+You can now try the example commands at http://napalm.readthedocs.org/en/latest/cli.html, using the test data files at https://github.com/napalm-automation/napalm-eos/tree/master/test/unit/eos::
+
+    # Get a sample test file (or copy/paste it from GitHub to new_good.conf):
+    $ wget https://raw.githubusercontent.com/napalm-automation/napalm-eos/master/test/unit/eos/new_good.conf
+
+    # dry run.
+    # (When prompted, the password is "vagrant")
+    $ cl_napalm_configure \
+      --user vagrant \
+      --vendor eos \
+      --strategy replace \
+      --optional_args 'port=12443' \
+      --dry-run \
+      new_good.conf localhost
+
+
+Try the other examples and commands given at http://napalm.readthedocs.org/en/latest/cli.html.
+
+Shutting down
+-------------
+
+    $ cd PROJECT_ROOT/docs/tutorials
```

Comment:
WALL OF TEXT ALERT!!! : )

Honestly, I would rather not merge yet. I see the value of what we are trying to do but I don't think we are there yet. I also think there is some confusion here. But that might be due to the latest changes in NAPALM or due to the lack of "common goals" for this PR.

First, some notes and clarifications regarding napalm.

> I would recommend keeping the tutorial in alongside the rest of the code, rather than as a separate repo

The repo `napalm` contains no code. It's only a placeholder for documentation and a wrapper to install all the drivers. The common code of napalm lives now in `napalm-base` and each driver has a dedicated repo, ie, `napalm-eos`, `napalm-ios`, etc... There was some discussion in the mailing list about this a month or two ago and the latest release notes has also some explanations around this and why we did it. In any case, what I meant by having a dedicated repo where the support files, not the content. However, I don't mind having the support files in the main `napalm` repo either, at the end of the day, this is one of the main reasons of this repo.

> the tutorial instructions should be tested alongside the code, to ensure that the instructions stay current. By "testing" here, I mean actual automated testing, and not relying on people to remember to test things when they change code. This isn't actually done here yet, we could backlog that idea.

That is not necessary. Two reasons:

1. The API is stable, that means that code written following napalm API should work, always, no matter what.
2. Each driver has it's own set of integration tests that are run automatically every time someone pushes new code. So we should find bugs and issues there without the need of testing the quickstart. I am not claiming the tests are perfect so feel free to comment if you think something could be improved. Bottom line here is that we already have automated testing covering this.

> "Release early, release often" is a good thing to follow, sometimes.

Agree, that was mostly the reason behind the latest refactoring. That and smaller PRs. However, features should be complete, releasing something that doesn't work might just frustrate people and make them lose confidence on the project so I am all for releasing quick and fast, as long as is complete. Specially if it's documentation. A common concern on OSS is that documentation is confusing or that you can't just follow it because you always find surprises.

> A colleague was asking for this quickstart material, if this gets into master and onto "readthedocs" it will help our team. :-)

But this is not about one particular's use case. It's about the community. Being a contributor has the advantage that you can focus on your particular problem but we should still aim to build something as generic as possible so everybody could get the benefit of it. I know that's probably not what you meant but I felt obliged to clarify anyway ; )

Back to the PR itself. Right now, I think that most of the content of this PR mostly duplicates work already done in the tutorials. So let me try to explain what I envision for this PR:

1. The target audience should be developers/network_engineers/users_in_general that heard about napalm and what to see it in action and maybe try it by themselves. The key word here is **users**. Anyone that might be considering napalm as their main interface to the network. They could be people looking into ansible, writing their own scripts or something else. But definitively not people trying to contribute to napalm. For that people we will have some other document.
2. The goal of the document should be to be able to copy & paste a few lines and see different things you can do with napalm:
   1. Change configuration
   1. Get diffs.
   1. Get LLDP neighbors/BGP/interfaces/whatever
   1. Test the CLI tool. 
3. Provide a list of resources to explore a bit further.

Right now, most of the stuff is covered on the tutorials, what is missing is:

1. How to quickly build a lab so you don't have to test on your own network.
2. Fixing the tutorials so you can actually copy & paste them.

In summary, my problem with the PR right now is that overlaps a lot with the tutorials. I would rather have this PR just telling you how to start a lab to move forward onto the rest of the tutorials and adapt the tutorials so they can be reproduced with the lab.

I hope it makes sense, otherwise, feel free to send me back a wall of text :P